torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -60,8 +60,23 @@ def __init__(
         super().__init__()
         self.range_constraints: Dict[sympy.Symbol, RangeConstraint] = range_constraints
         self.equality_constraints: List[Tuple[InputDim, InputDim]] = equality_constraints
+        # In some corner case, range constraint assertion could be added for
+        # the same symbol multiple times. For example:
+        # ```
+        # b = x.nonzero()
+        # constrain_as_value(b.shape[0], min=3, max=5)
+        # ```
+        # If x is a fake tensor, b could be fake tensor with size like (i0, 1). In this case,
+        # range constraint could be applied to `i0` at:
+        # 1. b = x.nonzero() - this is due to inside add_assertions it recursively applies
+        #    sym_size_cb until reaching scalar (then applies range constraint).
+        # 2. b.shape[0] passed to constrain_as_value(...).
+        # NOTE: cached_assertions will cause _AddRuntimeAssertionsForConstraintsPass stateful
+        # but only within each pass - it will always be reset at start of each pass.
+        self.cached_assertions: Set[Symbol] = set()
```

Comment:
What's the problem you're trying to solve?

1. Are duplicate assertions a problem you must solve in this diff?
2. If we want to deduplicate assertions, let's do it for all rather than just for inline constraints.

Reply:
Just to fix a failed test case https://github.com/pytorch/pytorch/blob/6cc0f1c20c2f87a6c7b0e4abd5419e5007920999/test/export/test_passes.py#L227.
But anyway, yeah it might be better to handle this separately and I guess we also need to handle case https://github.com/pytorch/pytorch/blob/93b0410eef57ebf038c12ed2fa1d4018a24096b7/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L101-L106.

I will revert this change and relax the test case for now and leave a TODO to handle duplication. cc @tugsbayasgalan  for awareness.

Reply:
What were the counts in that test before caching?

Reply:
I'm a bit uncomfortable not understanding what change in this diff caused the counts to change in that test, and therefore required caching. Specifically, are we doing something in this diff that *causes* the same assertion to be added twice, and can we avoid it there rather than doing caching here?

Reply:
So use the [test example](https://github.com/pytorch/pytorch/blob/a60f6dbe69b0cae2e27ac279cdf399c6c16650af/test/export/test_passes.py#L227), I think TLDR is that the second assertion comes from `constrain_as_value` survived AOT export.
More specifically:
Before the change and after AOT export, the logic is roughly like:
```
b = x.nonzero() # <- the assertion is for `b.shape[0]` here (it can smartly verify dimension of `b` because the recursion logic at https://fburl.com/bw1gihkx)
# constrain_as_value(b.shape[0], min=3, max=5) <- this statement is totally gone.
return b
```
After the change and after AOT export, the logic is roughly like:
```
b = x.nonzero() # <- the assertion is for `b.shape[0]` as before.
constrain_as_value(b.shape[0], min=3, max=5), min=2, max=5) # <- the new assertion is for `b.shape[0]` inside `constrain_as_value(...)`
```
Since _AddRuntimeAssertionsForConstraintsPass runs in interpretation mode, which means each statement is executed sequentially, so the the same assertion is added again when executing `b.shape[0]` which is the input of `constrain_as_value`. So the fix here is just to add a cache to avoid adding assertion for the same symbol twice.

Reply:
I see. @tugsbayasgalan @angelayi any thoughts here on whether this is expected behavior? How are we producing an assertion at `b = x.nonzero()`, and if so, how does it know the bounds are `(3,5)` without looking at the next line? Is it because in an earlier pass we discovered the constraint to add, so in the next pass we're adding it again? If so, can't we check the shape env or something to avoid adding it again?

To be clear, I have nothing against doing caching to avoid this duplication. What I'm actually worried about is that this is hiding some bug, or at least some ugliness, in our logic. Please let me know if I'm wrong.

Reply:
Also did another sanity check, for the follow logic **before the PR**:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        b = x.nonzero()
        constrain_as_value(b.shape[0], min=3, max=5)
        z = b.shape[0] + 1
        return z

mod = M()
x = torch.tensor([2, 1, 2, 3, 5, 0])
ep = export(mod, (x,), constraints=[dynamic_dim(x, 0) >= 2])

print(count_call_function(ep.graph, torch.ops.aten._assert_async.msg)) # this will be 4 as well.
```
So I think this verifies this is not a new behavior introduced by this PR :p

Reply:
Could you paste the whole graph before your change here?

Reply:
So the graph (before my change) for the code snippets I've posted above is:
```graph():
    %arg0_1 : [#users=2] = placeholder[target=arg0_1]
    %sym_size_int : [#users=0] = call_function[target=torch.ops.aten.sym_size.int](args = (%arg0_1, 0), kwargs = {})
    %nonzero_default : [#users=2] = call_function[target=torch.ops.aten.nonzero.default](args = (%arg0_1,), kwargs = {})
    %sym_size_int_1 : [#users=2] = call_function[target=torch.ops.aten.sym_size.int](args = (%nonzero_default, 0), kwargs = {})
    %ge : [#users=1] = call_function[target=operator.ge](args = (%sym_size_int_1, 3), kwargs = {})
    %scalar_tensor_default : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%ge,), kwargs = {})
    %_assert_async_msg : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default, nonzero_default.shape[0] is outside of inline constraint [3, 5].), kwargs = {})
    %le : [#users=1] = call_function[target=operator.le](args = (%sym_size_int_1, 5), kwargs = {})
    %scalar_tensor_default_1 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%le,), kwargs = {})
    %_assert_async_msg_1 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_1, nonzero_default.shape[0] is outside of inline constraint [3, 5].), kwargs = {})
    %sym_size : [#users=3] = call_function[target=torch.ops.aten.sym_size](args = (%nonzero_default, 0), kwargs = {})
    %ge_1 : [#users=1] = call_function[target=operator.ge](args = (%sym_size, 3), kwargs = {})
    %scalar_tensor_default_2 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%ge_1,), kwargs = {})
    %_assert_async_msg_2 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_2, sym_size is outside of inline constraint [3, 5].), kwargs = {})
    %le_1 : [#users=1] = call_function[target=operator.le](args = (%sym_size, 5), kwargs = {})
    %scalar_tensor_default_3 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%le_1,), kwargs = {})
    %_assert_async_msg_3 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_3, sym_size is outside of inline constraint [3, 5].), kwargs = {})
    %add : [#users=1] = call_function[target=operator.add](args = (%sym_size, 1), kwargs = {})
    return (add,)
```

Actually I feel like there is another issue? Should there be a constrains for `dynamic_dim(x, 0) >= 2`? Maybe I missed it or misunderstand things :p.

Reply:
Hmm, yes seems like a bug 🤔

Reply:
Oh wait sorry, it's not a bug, because we always want the shapes to be >= 2.